### PR TITLE
Base validation

### PR DESCRIPTION
**Pull request by IgorZiegler**
- Add base validation which parse all validation rules in afterrender function and compiled it in validate function.
 - Add rules registration module which returns new parser with registered functions and properties
 - Add examples of validation rules in text.json 
 - Added base popover logic
 - If you want to add popover to specific element just redefine get popoverEl method for your block
 - Redefine popoverEl Method for input

### CODE REVIEW

File: frontend/blocks/base/base.js [hunk 1]
```diff
@@ -19,12 +20,57 @@ export default class Block {
...
+    }
+
+    return Object.assign({fieldName: this.name}, parent.value);
   }
 
   validate() {
+    if (this.compiledRules && !this.checkCompiledRules()) {
+      return false;
+    }
+    this.popover('hide');
```
Comment (IgorZiegler) [inline frontend/blocks/base/base.js]: Romove this hide =)

File: frontend/blocks/base/base.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ import $ from 'jquery';
 import cloneDeep from 'lodash/cloneDeep';
 import defaults from 'lodash/defaults';
 import uniqueId from 'lodash/uniqueId';
+import parser from '../../modules/rules_registrator';
```
Comment (102) [inline frontend/blocks/base/base.js]: `rules_registrar`

File: frontend/blocks/base/base.js (hunk 1 above)
Comment (102) [inline frontend/blocks/base/base.js]: Why do we want to `hide` if we'll `destroy` it in next step?

File: frontend/blocks/base/base.js [hunk 3]
```diff
@@ -19,12 +20,57 @@ export default class Block {
...
+    }
+    this.popover('hide');
+    this.popover('destroy');
+    this.popoverExist = false;
+
     return true;
   }
 
+  checkCompiledRules() {
+    return !this.compiledRules.some((rule) => {
```
Comment (102) [inline frontend/blocks/base/base.js]: extra parenthesis

File: frontend/blocks/base/base.js [hunk 4]
```diff
@@ -19,12 +20,57 @@ export default class Block {
   }
 
   afterRender() {
+    if (this.config.validationRules && Array.isArray(this.config.validationRules)) {
```
Comment (maxmalov) [inline frontend/blocks/base/base.js]: use early returns

File: frontend/blocks/base/base.js [hunk 5]
```diff
@@ -19,12 +20,57 @@ export default class Block {
...
+  }
+
+  get context() {
+    let parent = this.parent;
+
+    while (parent && null !== parent.parent) {
+      parent = parent.parent;
+    }
+
+    return Object.assign({fieldName: this.name}, parent.value);
```
Comment (maxmalov) [inline frontend/blocks/base/base.js]: missing spaces after and before `{ }`

File: frontend/blocks/base/base.js (hunk 3 above)
Comment (maxmalov) [inline frontend/blocks/base/base.js]: `rule => `

File: frontend/modules/rules_registrator.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+const parser = new Parser();
+
+parser.registerProperty('обязательное', scope => {
+  const value = scope[scope.fieldName];
+
+  return value && '' !== value && false !== value && null !== value;
+});
+
+parser.registerProperty('длинна', scope => {
```
Comment (mvahrameev) [inline frontend/modules/rules_registrator.js]: *Длина

File: frontend/blocks/base/base.js [hunk 7]
```diff
@@ -60,14 +106,42 @@ export default class Block {
     return this._id;
   }
 
+  get name() {
+    return this.config.name;
+  }
+
+  getErrorMessage() {
+    return () => this.errorMessage;
```
Comment (maxmalov) [inline frontend/blocks/base/base.js]: this isn't a message, it is a function!

Comment (maxmalov) [inline frontend/blocks/base/base.js]: and whats the difference with https://github.com/hwdtech/smartforms/pull/71/files#diff-34b111a15cdd4629a6991f7f8af4f736R117?

File: public/examples/forms/text.json [hunk 8]
```diff
@@ -4,12 +4,30 @@
...
+      "block": "text",
+      "name": "accountNumber1",
+      "label": "Лицевой счёт с двойным правилом",
       "textAlign": "center",
-      "rightMark": "<span class=\"glyphicon glyphicon-ruble\" aria-hidden=\"true\" style=\"font-size: 11px\"><\/span>"
+      "value": "1234 9999999",
+      "rightMark": "<span class=\"glyphicon glyphicon-ruble\" aria-hidden=\"true\" style=\"font-size: 11px\"><\/span>",
+      "validationRules": [{
+        "rule": "обязательное && длинна > 5",
+        "errorMessage": "Поле должно быть заполнено длина поля должна быть больше 5"
```
Comment (102) [inline public/examples/forms/text.json]: `"Поле должно быть заполнено длина поля должна быть больше 5"`

:confused:

File: frontend/blocks/base/base.js (hunk 3 above)
Comment (abukurov) [inline frontend/blocks/base/base.js]:
```js
const invalidRule = this.compiledRules.find(rule => rule.compiledRule.eval(this.context));
if (invalidRule) {
  // show error
}
```

File: frontend/blocks/base/base.js [hunk 9]
```diff
@@ -60,14 +106,42 @@ export default class Block {
...
     });
   }
 
-  popover(...args) {
-    this.el.popover(...args);
+  get popoverEl() {
+    return this.el;
+  }
+
+  get popoverExist() {
```
Comment (maxmalov) [inline frontend/blocks/base/base.js]: this `get popoverExist() {}` and `set popoverExists() {}` are redundant

File: public/examples/forms/text.json (hunk 8 above)
Comment (mvahrameev) [inline public/examples/forms/text.json]: Missing comma

File: frontend/blocks/input/input.js [hunk 10]
```diff
@@ -8,6 +8,10 @@ export default class Input extends Base {
     return template;
   }
 
+  get popoverEl() {
+    return this.el.find('input');
```
Comment (maxmalov) [inline frontend/blocks/input/input.js]: every time it will call `find`, it isn't fast

File: frontend/modules/rules_registrator.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import Parser from 'morph-expressions';
+
+const parser = new Parser();
+
+parser.registerProperty('обязательное', scope => {
+  const value = scope[scope.fieldName];
+
+  return value && '' !== value && false !== value && null !== value;
```
Comment (maxmalov) [inline frontend/modules/rules_registrator.js]: `lodash.isEmpty`?

File: frontend/modules/rules_registrator.js (hunk 6 above)
Comment (maxmalov) [inline frontend/modules/rules_registrator.js]: длина

File: frontend/blocks/base/base.js [hunk 12]
```diff
@@ -19,12 +20,57 @@ export default class Block {
   }
 
   afterRender() {
+    if (this.config.validationRules && Array.isArray(this.config.validationRules)) {
+      this.compiledRules = this.config.validationRules.map(rule => {
```
Comment (abukurov) [inline frontend/blocks/base/base.js]: ` this.config.validationRules.map(rule => Object.assign(...))`

File: package.json [hunk 13]
```diff
@@ -46,16 +46,14 @@
     "jquery.formstyler": "^1.7.5-repackage",
     "jquery.maskedinput": "^1.4.1",
     "lodash": "^4.12.0",
-    "morph-expressions": "git+https://github.com/abukurov/morph-expressions.git#develop",
+    "morph-expressions": "0.0.4",
```
Comment (maxmalov) [inline package.json]: why is this a dev dependency?

**Comment by maxmalov** [comment]
Where are unit tests?

File: frontend/blocks/base/base.js (hunk 2 above)
Comment (abukurov) [inline frontend/blocks/base/base.js]: parser

File: public/examples/forms/text.json [hunk 14]
```diff
@@ -4,12 +4,30 @@
...
+      "value": "1234 9999999",
       "mask": "9999 9999999",
       "maskPlaceholder": "_",
-      "value": "1234 56789123",
+      "rightMark": "<span class=\"glyphicon glyphicon-ruble\" aria-hidden=\"true\" style=\"font-size: 11px\"><\/span>",
+      "validationRules": [{
+          "rule": "обязательное",
+          "errorMessage": "Поле должно быть заполнено"
+      }, {
+        "rule": "длинна > 5",
```
Comment (mvahrameev) [inline public/examples/forms/text.json]: Длина

File: frontend/blocks/base/base.js (hunk 4 above)
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: Array.isArray is enough

**Comment by IgorZiegler** [comment]
@maxmalov This is [next issue](https://github.com/hwdtech/smartforms/issues/72)

File: frontend/blocks/base/base.js [hunk 15]
```diff
@@ -3,6 +3,7 @@ import $ from 'jquery';
 import cloneDeep from 'lodash/cloneDeep';
 import defaults from 'lodash/defaults';
 import uniqueId from 'lodash/uniqueId';
+import parser from '../../service/parser';
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: Maybe services?

File: frontend/blocks/base/base.js [hunk 16]
```diff
@@ -19,12 +20,54 @@ export default class Block {
...
+    if (this.compiledRules && !this.checkCompiledRules()) {
+      return false;
+    }
+    this.popover('destroy');
+    this._popoverExist = false;
+
     return true;
   }
 
+  checkCompiledRules() {
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: It seems like this function should return bool result

File: frontend/blocks/base/base.js [hunk 17]
```diff
@@ -19,12 +20,54 @@ export default class Block {
...
+  }
+
+  get context() {
+    let parent = this.parent;
+
+    while (parent && null !== parent.parent) {
+      parent = parent.parent;
+    }
+
+    return Object.assign({ fieldName: this.name }, parent.value);
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: What about situation when parent is undefined?

File: frontend/blocks/base/base.js [hunk 18]
```diff
@@ -19,12 +20,54 @@ export default class Block {
   }
 
   afterRender() {
+    if (Array.isArray(this.config.validationRules)) {
+      this.compiledRules = this.config.validationRules.map(rule => Object.assign({
+        compiledRule: parser.parse(rule.rule)
+      }, rule));
+
+      return;
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: Why?

File: frontend/service/parser.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+parser.registerProperty('обязательное', scope => {
+  const value = scope[scope.fieldName];
+
+  return !isNil(value) && '' !== value;
+});
+
+parser.registerProperty('длина', scope => {
+  const value = scope[scope.fieldName];
+
+  return value.length;
```
Comment (mvahrameev) [inline frontend/service/parser.js]: Can `value` be undefined?

File: frontend/blocks/base/base.js [hunk 20]
```diff
@@ -19,12 +20,58 @@ export default class Block {
...
+    this.popover('destroy');
+    this._popoverExist = false;
+
+    return true;
+  }
+
+  checkCompiledRules() {
+    const invalidRule = this.compiledRules.find(rule => !rule.compiledRule.eval(this.context));
+
+    if (invalidRule) {
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: early return

File: public/examples/forms/text.json [hunk 21]
```diff
@@ -4,12 +4,30 @@
...
       "mask": "9999 9999999",
       "maskPlaceholder": "_",
-      "value": "1234 56789123",
+      "rightMark": "<span class=\"glyphicon glyphicon-ruble\" aria-hidden=\"true\" style=\"font-size: 11px\"><\/span>",
+      "validationRules": [{
+          "rule": "обязательное",
+          "errorMessage": "Поле должно быть заполнено"
+      }, {
+        "rule": "длина > 5",
+        "errorMessage": "длина поля должна быть больше 5"
```
Comment (mvahrameev) [inline public/examples/forms/text.json]: Captilize first letter

File: frontend/blocks/base/base.js (hunk 15 above)
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: I think you will have more than one service...

File: public/examples/forms/text.json (hunk 21 above)
Comment (IgorZiegler) [inline public/examples/forms/text.json]: Capitalize*

File: frontend/blocks/base/base.js (hunk 20 above)
Comment (IgorZiegler) [inline frontend/blocks/base/base.js]: fixed

File: frontend/blocks/base/base.js [hunk 22]
```diff
@@ -19,12 +20,58 @@ export default class Block {
...
+
     return true;
   }
 
+  showErrorMessage() {
+    if (!this._popoverExist) {
+      this.popover({
+        placement: this.config.popoverPlacement || 'top',
+        content: this.getErrorMessageFn()
+      });
```
Comment (mvahrameev) [inline frontend/blocks/base/base.js]: 'this._popoverExist = true;'

**Pull request merged**